src/FSharp.Data.GraphQL.Server.AspNetCore/Giraffe/HttpHandlers.fs

```diff
@@ -1,155 +1,268 @@
-                CancellationTokenSource
-                    .CreateLinkedTokenSource(cancellationToken, ctx.RequestAborted)
-                    .Token
-            if cancellationToken.IsCancellationRequested then
-                return (fun _ -> None) ctx
-            else
-                let options = ctx.RequestServices.GetRequiredService<GraphQLOptions<'Root>> ()
-                let executor = options.SchemaExecutor
-                let rootFactory = options.RootFactory
-                let serializerOptions = options.SerializerOptions
-                let deserializeGraphQLRequest () = task {
-                    try
-                        let! deserialized = JsonSerializer.DeserializeAsync<GraphQLRequest> (ctx.Request.Body, serializerOptions)
-                        return Ok deserialized
-                    with :? GraphQLException as ex ->
-                        logger.LogError (``exception`` = ex, message = "Error while deserializing request.")
-                        return Result.Error [ $"%s{ex.Message}\n%s{ex.ToString ()}" ]
-                }
 
-                let applyPlanExecutionResult (result : GQLExecutionResult) = task {
-                    let gqlResponse =
-                        match result.Content with
-                        | Direct (data, errs) -> GQLResponse.Direct (result.DocumentId, data, errs)
-                        | RequestError (problemDetailsList) -> GQLResponse.RequestError (result.DocumentId, problemDetailsList)
-                        | _ ->
-                            GQLResponse.RequestError (
-                                result.DocumentId,
-                                [
-                                    GQLProblemDetails.Create ("subscriptions are not supported here (use the websocket endpoint instead).")
-                                ]
+    let rec private moduleType = getModuleType <@ moduleType @>
+
+    let ofTaskIResult ctx (taskRes: Task<IResult>) : HttpFuncResult = task {
+        let! res = taskRes
+        do! res.ExecuteAsync(ctx)
+        return Some ctx
+    }
+
+    let ofTaskIResult2 ctx (taskRes: Task<Result<IResult, IResult>>) : HttpFuncResult =
+        taskRes
+        |> TaskResult.defaultWith id
+        |> ofTaskIResult ctx
+
+    /// Set CORS to allow external servers (React samples) to call this API
+    let setCorsHeaders : HttpHandler =
+        setHttpHeader "Access-Control-Allow-Origin" "*"
+        >=> setHttpHeader "Access-Control-Allow-Headers" "content-type"
+
+    let private handleGraphQL<'Root> (next : HttpFunc) (ctx : HttpContext) =
+        let sp = ctx.RequestServices
+
+        let logger = sp.CreateLogger moduleType
+
+        let options = sp.GetRequiredService<GraphQLOptions<'Root>>()
+
+        let toResponse { DocumentId = documentId; Content = content; Metadata = metadata } =
+
+            let serializeIdented value =
+                let jsonSerializerOptions = options.GetSerializerOptionsIdented()
```

Comment:
Ok, now I see where you're using `.GetSerializerOptionsIdented()`. I'm concerned about [performance penalties](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/configure-options?pivots=dotnet-8-0) coming from creating a new instance here every time.
Besides that, since these jsonSerializerOptions are specific to the purpose of logging, actually they should be managed completely separate from the other options IMHO. That means: no copying the jsonSerializerOptions which the application code (I mean: the code that uses FSharp.Data.GraphQL) can customize, but having an internal, not customizable jsonSerializerOptions instantiated once and used solely for the purpose of logging.

Reply:
I agree, propose your solution

Reply:
My intent was to have the same options as the default but with indentation